src/fs/9p.c

```diff
@@ -0,0 +1,813 @@
+        p9_debug("  file %p, datasync %d\n", f, datasync);
+        fs_status fss;
+        filesystem_lock(&fs->fs);
+        if (f)
+            fss = v9p_fsync(transport, f->dentry->fid, datasync);
+        else
+            list_foreach(&fs->fsfiles, e) {
+                f = struct_from_list(e, p9_fsfile, l);
+                fss = v9p_fsync(transport, f->dentry->fid, datasync);
+                if (fss != FS_STATUS_OK)
+                    break;
+            }
+        filesystem_unlock(&fs->fs);
+        if (fss != FS_STATUS_OK)
+            s = timm("result", "fsync failed (%d)", fss);
+    }
+    async_apply_status_handler(bound(completion), s);
+    closure_finish();
+}
+
+closure_function(2, 2, boolean, p9_dir_cleanup,
+                 p9fs, fs, tuple, other_c,
+                 value, k, value, v)
+{
+    p9fs fs = bound(fs);
+    tuple other_c = bound(other_c);
+    if (!other_c || !get_tuple(other_c, k)) {
+        p9_dentry dentry = p9_get_dentry_from_md(fs, v);
+        if (dentry) {
+            fs_notify_release(dentry->md, false);
+            p9_dentry_delete(fs, dentry);
+        } else {
+            deallocate_value(v);
+        }
+    }
+    return true;
+}
+
+static fs_status p9_readdir(p9fs fs, u32 fid, tuple md)
+{
+    p9_debug("readdir: fid %d, md %p\n", fid, md);
+    u64 qid;
+    u32 iounit;
+    fs_status s = v9p_lopen(fs->transport, fid, O_RDONLY, &qid, &iounit);
+    if (s == FS_STATUS_INVAL)   /* this happens if fid has been already opened */
+        s = FS_STATUS_OK;
+    else if (s != FS_STATUS_OK)
+        return s;
+    const int iobuf_size = PAGESIZE;
+    u8 *buf = v9p_get_iobuf(fs->transport, iobuf_size);
+    if (buf == INVALID_ADDRESS)
+        return FS_STATUS_NOMEM;
+    tuple old_c = children(md);
+    tuple new_c = allocate_tuple();
+    u64 offset = 0;
+    u32 count;
+    p9_readdir_entry entry;
+    do {
+        s = v9p_readdir(fs->transport, fid, offset, buf, iobuf_size, &count);
+        if (s != FS_STATUS_OK)
+            break;
+        u64 buf_offset = 0;
+        while (count > buf_offset + sizeof(*entry)) {
+            entry = (void *)buf + buf_offset;
+            if ((entry->name.length > NAME_MAX) ||
+                (buf_offset + sizeof(*entry) + entry->name.length > count)) {
+                s = FS_STATUS_IOERR;
+                break;
+            }
+            if (p9_strcmp(&entry->name, ".") && p9_strcmp(&entry->name, "..")) {
+                symbol name_sym = p9_sym(&entry->name);
+                tuple t;
+                if (old_c)
+                    t = get_tuple(old_c, name_sym);
+                else
+                    t = 0;
+                p9_dentry dentry;
+                if (t) {
+                    dentry = p9_get_dentry_from_md(fs, t);
+                } else {
+                    t = allocate_tuple();
+                    dentry = 0;
+                }
+                p9_debug("  dir entry '%s', type 0x%x, md %p\n", name, entry->type, t);
```

Comment:
'name' no longer exists, have to use name_sym and a different format specifier

Reply:
Thanks, I replaced 'name' with 'symbol_string(name_sym)', using %b as format specifier.